File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
    productExceptSelf(nums) {
        let left = []
        let prod = 1
        for (let i = 0; i < nums.length; i ++) {
            left.push(prod)
            prod = prod * nums[i]
        }


        prod = 1
        let right = []
        for (let i = nums.length - 1; i >= 0; i --) {
            right.unshift(prod)
            prod = prod * nums[i]
            
        }

        let result = []
        for (let j = 0; j < nums.length; j++){
            result.push(left[j] * right[j])
        }

        return result
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-0.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()

        result = []

        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i - 1]:
                continue  # Skip duplicate values
            if nums[i] > 0:
                break
            
            l, r = i + 1, len(nums) - 1

            while l < r:
                curSum = nums[i] + nums[l] + nums[r]

                if curSum  > 0 :
                    r -= 1
                elif curSum < 0:
                    l += 1
                else:
                    result.append([nums[i], nums[l], nums[r]])
                    r -= 1
                    l += 1
                    # Skip duplicate values on the left
                    while l < r and nums[l] == nums[l - 1]:
                        l += 1

        return result
